tensorflow/contrib/tensorrt/convert/convert_graph.cc

```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+        LOG(ERROR) << "Failed to construct calibration storage";
+      }
+      break;
+    }
+    default: {
+      return tensorflow::errors::OutOfRange("Unknown precision mode");
+    }
+  }
+  tensorflow::Status status;
+  tensorflow::Node* engine_node = nullptr;
+  tensorflow::NodeDefBuilder node_builder(info.engine_name, "TRTEngineOp");
+  if (!info.device.empty()) {
+    node_builder.Device(info.device);
+  }
+  if (VLOG_IS_ON(1)) {
+    string ins(info.engine_name);
+    for (const auto& ii : inputs) {
+      StrAppend(&ins, ii.node, ":", ii.index, " ");
+    }
+    VLOG(1) << ins;
+  }
+  node_builder.Input(inputs);
+  if (info.engine_type == EngineInfo::EngineType::TRTStatic) {
+    if (info.cached_engine_batches.size()) {
+      LOG(WARNING) << "Cached engine batches are ignored for static engines";
+    }
+  }
+  status = node_builder.Attr("input_shapes", input_shapes)
+               .Attr("output_shapes", out_shapes)
+               .Attr("static_engine",
+                     info.engine_type == EngineInfo::EngineType::TRTStatic)
+               .Attr("segment_funcdef_name",
+                     StrCat(info.engine_name, "_native_segment"))
+               .Attr("serialized_segment", segment_string)
+               .Attr("calibration_data", "")
+               .Attr("max_cached_engines_count", info.maximum_cached_engines)
+               .Attr("cached_engine_batches", {max_batch_size})
+               .Attr("workspace_size_bytes", info.max_workspace_size_bytes)
+               .Attr("precision_mode", prec_string)
+               .Attr("OutT", out_types)
+               .Finalize(trtNode);
+  if (!status.ok()) {
+    LOG(ERROR) << "Node construction failed with" << status;
+    return status;
+  }
+  VLOG(1) << "Adding TRTEngine " << info.engine_name << " to graph";
+  engine_node = graph->AddNode(*trtNode, &status);
+  if (!status.ok()) {
+    LOG(ERROR) << "Adding node failed " << status;
+    return status;
+  }
+
+  for (auto& conn : info.connections) {
+    if (conn.is_input_edge) continue;
+    VLOG(1) << " Updating DBG " << engine_node->name() << " out_port "
+            << conn.port_number << " out_id " << conn.outside_id
+            << " name=" << conn.outside_node_name;
+    auto dst_node = graph->FindNodeId(conn.outside_id);
+    if (!dst_node) {  // node removed skip.
+      continue;
+    }
+    VLOG(1) << "Updating " << engine_node->name() << ":" << conn.port_number
+            << " to " << dst_node->name() << ":" << conn.outside_port;
+    status = graph->UpdateEdge(engine_node, conn.port_number, dst_node,
+                               conn.outside_port);
+    if (!status.ok()) {
+      LOG(ERROR) << "Edge update failed " << engine_node->name() << ":"
+                 << conn.port_number << " -> " << dst_node->name() << ":"
+                 << conn.outside_port << " status= " << status;
+    }
+  }
+  return status;
+}
+
+// Function to construct a funcdef from the segment and add it to the graph.
```

Comment:
Please move the comment to the header.

Reply:
This is an internal method. Not exposed in header.